Create ultralytics_yolov5.md

glenn-jocher opened the pull request:
🚀 This PR submits Ultralytics YOLOv5 for consideration for inclusion into PyTorch Hub. 

We've verified correct operations of the tutorials presented in this PR, and also added a PyTorch Hub section directly to our README at https://github.com/ultralytics/yolov5#pytorch-hub to improve awareness understanding of this feature.

We've streamlined YOLOv5 workflows to allow for an absolute minimum viable code to run full inference on PIL or OpenCV images (including batched inference) with automatic source handling, image reshaping/padding, NMS and post-processing all included and handled automatically:

```python
import torch
from PIL import Image

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True).fuse().eval()  # yolov5s.pt
model = model.autoshape()  # for autoshaping of PIL/cv2/np inputs and NMS

# Images
torch.hub.download_url_to_file('https://github.com/ultralytics/yolov5/releases/download/v1.0/zidane.jpg', 'zidane.jpg')
img = Image.open('zidane.jpg')  # PIL image

# Inference
prediction = model(img)  # includes NMS
```

Let us know if any changes are required please. All feedback is welcome!

glenn-jocher commented:
circleci checks are failing due to an unrelated problem that seems to originate with huggingface. I see an open PR on the topic as well https://github.com/pytorch/hub/pull/161. Netlify preview looks good.

```bash
Running pytorch example in huggingface_pytorch-transformers.md
Downloading: "https://github.com/huggingface/pytorch-transformers/archive/master.zip" to /home/circleci/.cache/torch/hub/master.zip
Traceback (most recent call last):
  File "temp.py", line 2, in <module>
    tokenizer = torch.hub.load('huggingface/pytorch-transformers', 'tokenizer', 'bert-base-cased')
  File "/home/circleci/miniconda3/lib/python3.7/site-packages/torch/hub.py", line 349, in load
    hub_module = import_module(MODULE_HUBCONF, repo_dir + '/' + MODULE_HUBCONF)
  File "/home/circleci/miniconda3/lib/python3.7/site-packages/torch/hub.py", line 71, in import_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/hubconf.py", line 8, in <module>
    from transformers import (
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/__init__.py", line 135, in <module>
    from .pipelines import (
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/pipelines.py", line 38, in <module>
    from .tokenization_auto import AutoTokenizer
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_auto.py", line 118, in <module>
    from .tokenization_albert_fast import AlbertTokenizerFast
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_albert_fast.py", line 23, in <module>
    from .tokenization_utils_fast import PreTrainedTokenizerFast
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_utils_fast.py", line 30, in <module>
    from .convert_slow_tokenizer import convert_slow_tokenizer
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/convert_slow_tokenizer.py", line 27, in <module>
    from transformers.utils import sentencepiece_model_pb2 as model
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/utils/sentencepiece_model_pb2.py", line 9, in <module>
    from google.protobuf import descriptor as _descriptor
ModuleNotFoundError: No module named 'google'
```

ailzhang commented:
@glenn-jocher Thanks for contributing to hub! I've fixed the dependency issue in a separate PR, would you mind rebasing on top of master and let the CI run again? Thanks!

glenn-jocher commented:
@ailzhang thanks! I've rebased and committed bug fixes. CI checks are all green now. Netlify preview looks good, though I noticed the preview is missing the github star count.

ailzhang commented:
gentlely ping @soumith on reviewing ;)

glenn-jocher commented:
CI checks are failing on missing CLA and an apparent unrelated circleci check. Corporate CLA is now signed for Ultralytics LLC, with GitHub username glenn-jocher added.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+accelerator: cuda-optional
+---
+
+🚀  This guide explains how to load YOLOv5 from PyTorch Hub [https://pytorch.org/hub/](https://pytorch.org/hub/)
```
remove this line. seems redundant

glenn-jocher replied:
Done.

soumith commented on ultralytics_yolov5.md:
````diff
@@ -0,0 +1,180 @@
+$ pip install -U opencv-python pillow pyyaml tqdm  # install dependencies
+```
+
+Note that cloning the [https://github.com/ultralytics/yolov5](https://github.com/ultralytics/yolov5) repository is **not** required.
````
remove this line. it's redundant for pytorch hub models

glenn-jocher replied:
Done.

soumith commented on ultralytics_yolov5.md:
````diff
@@ -0,0 +1,180 @@
+    print(str + 'Done.')
+```
+
+<img src="https://user-images.githubusercontent.com/26833433/96368758-114c3a00-1156-11eb-9ede-a03f9aff42e4.jpg" width="550">
````
remove these two inlined images. they are not needed, as when the user opens the hub page in a notebook and runs them, they will see the output

glenn-jocher replied:
Done.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+img = Image.open('zidane.jpg')  # PIL image
+
+# Inference
+prediction = model(img)  # includes NMS
```
describe the output `prediction` here in a sentence.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+with torch.no_grad():
+    prediction = model(imgs, size=640)  # includes NMS
+
+# Plot
```
break out the plotting into a separate code block

glenn-jocher replied:
Done.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+
+<img src="https://user-images.githubusercontent.com/26833433/97121810-54dc1080-1721-11eb-8c25-451ccb1452e4.jpg" width="350">
+
+For all inference options see YOLOv5 [`autoShape()` forward method](https://github.com/ultralytics/yolov5/blob/453acdec67202e15b772370d31eb196d844f2892/models/common.py#L129-L136):
```
The PyTorch Hub card is not meant to be a complete reference. It is meant to be a card for gentle introduction.

The card has already gotten too long.
I suggest that from line 96 onwards, you remove everything, and instead add a pointer to your repository or documentation that showcases more options.

glenn-jocher replied:
Understood, update done.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+
+## Load YOLOv5 From PyTorch Hub
+
+### Simple Example
```
let's aim to have just one example for the whole card, a simple example showcasing plotting as well.

So, please merge the Simple Example and Detailed Example sections into a single concise section.

glenn-jocher replied:
Good point, update done.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+
+<img width="800" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/97808084-edfcb100-1c64-11eb-83eb-ffed43a0859f.png">
+
+YOLOv5 is an architecture for compound-scaled object detection models trained on the COCO 2017 dataset, and includes built-in functionality for Test Time Augmentation (TTA), Model Ensembling, Rectangular Inference, Hyperparameter Evolution and export to a variety of common formats such as ONNX, CoreML and TFLite. YOLOv5 also features a free **iOS App** that demos real-time 30 FPS inference of all ...
```
you can summarize this description in a shorter form right before the `Load YOLOv5 From PyTorch Hub` section

glenn-jocher replied:
Yes, it's been shortened now.

soumith commented on ultralytics_yolov5.md:
```diff
@@ -0,0 +1,180 @@
+- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)
+
+
+## Citation
```
keep the Citation and Contact sections

glenn-jocher replied:
Ok!

soumith reviewed (changes requested):
Thanks for the contribution. It's really high quality and great.

The PyTorch Hub cards are meant to be short and concise and a gentle introduction.
With that in mind, I've requested some changes, with detailed comments in-line.

glenn-jocher commented:
@ailzhang @soumith thanks for the detailed review! I've gone ahead and implemented all requested changes.

**I've converted the PR into a draft temporarily** though as we are updating our torch hub functionality and want to include these updates in this PR. The updates will allow for much simplified plotting and printing of inference results, and will allow us to further simplify the card example. I think this may be ready in about a week's time. I will update here when that is complete.

soumith commented:
sounds good

glenn-jocher commented:
@soumith the YOLOv5 PyTorch Hub updates are complete now. We've updated our inference in https://github.com/ultralytics/yolov5/pull/1415 to make everything super easy, no more messy plotting code:

```
# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True).autoshape()  # for PIL/cv2/np inputs and NMS

# Images
img = Image.open('zidane.jpg')  # PIL image

# Inference
results = model(img)  # includes NMS

# View
results.show()  # .show() results, .save() jpgs, or .print() to screen
```

CI appears to be failing due to an unrelated huggingface_pytorch-transformers.md issue.

soumith reviewed (approved):
looks great! Thanks a lot for the contribution to PyTorch Hub.

soumith commented:
the PR is still marked as Draft, please fix that to a final PR so that I can merge it.

Once I merge the PR, it will be live on the PyTorch website after 24 hours.

glenn-jocher commented:
@soumith awesome, thank you!! :)
